### PR TITLE
Add `data:install:migrations` rake task

### PR DESCRIPTION
**Pull request by bolshakov**
It is a sibling of `db:install:migrations` rake task and
inteded to copy migrations from Rails Engine to application.

### CODE REVIEW

File: lib/tasks/data_migrations.rake [hunk 1]
```diff
@@ -1,4 +1,5 @@
 require 'rake'
+require 'rails-data-migrations'
```
Comment (serggl) [inline lib/tasks/data_migrations.rake:2]: I suspect this line came from here https://github.com/OffgridElectric/rails-data-migrations/pull/3
Well If I comment this line, all tests passes for me (`appraisal rake`), so I would still avoid this line. 
In my understanding this is something that bundler should do

File: Rakefile [hunk 2]
```diff
@@ -3,4 +3,14 @@ require 'rspec/core/rake_task'
 
 RSpec::Core::RakeTask.new(:spec)
 
-task :default => :spec
+task default: [:setup, :spec]
+
+desc 'Creates a test rails app for the specs to run against'
+task :setup do
+  source_path = 'spec/dummy'
```
Comment (serggl) [inline Rakefile:10]: I dont like `dummy` file/folder names. Can it have more meaningful names?

File: Rakefile [hunk 3]
```diff
@@ -3,4 +3,14 @@ require 'rspec/core/rake_task'
...
+  template = 'spec/support/dummy_app_template.rb'
+  system("rm -rf #{source_path}")
+  system(<<-BASH)
+    bundle exec rails new #{source_path} -m #{template} --skip-spring --skip-git --skip-turbolinks --skip-test-unit
```
Comment (serggl) [inline Rakefile]: `-T` seem more reasonable option here instead of `--skip-test-unit`

File: spec/tasks/data_migraitons_spec.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
+require 'spec_helper'
```
Comment (serggl) [inline spec/tasks/data_migraitons_spec.rb]: there is a typo in filename. Also same file already exists in spec folder, so I suggest it should be renamed

**Review by serggl — COMMENTED** [review]
@bolshakov there are couple of notices, see above.
Also tests became a little slow, any chance to make them run faster?

File: lib/tasks/data_migrations.rake (hunk 1 above)
Comment (bolshakov) [reply lib/tasks/data_migrations.rake:2]: Actually it does not if you run generator within rails engine.
I've factored a demo https://github.com/bolshakov/rails-data-migration-require-with-engine-demo

To reproduce the error:

1. `git clone https://github.com/bolshakov/rails-data-migration-require-with-engine-demo`
2. `bundle install`
3. `bin/rails g data_migration Foo`

And it fails with error:

```ruby
/usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/rails-data-migrations-1.0.4/lib/generators/data_migration_generator.rb:7:in `create_migration_file': uninitialized constant DataMigrationGenerator::RailsDataMigrations (NameError)
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/thor-0.19.4/lib/thor/command.rb:27:in `run'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/thor-0.19.4/lib/thor/invocation.rb:126:in `invoke_command'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/thor-0.19.4/lib/thor/invocation.rb:133:in `block in invoke_all'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/thor-0.19.4/lib/thor/invocation.rb:133:in `each'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/thor-0.19.4/lib/thor/invocation.rb:133:in `map'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/thor-0.19.4/lib/thor/invocation.rb:133:in `invoke_all'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/thor-0.19.4/lib/thor/group.rb:232:in `dispatch'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/thor-0.19.4/lib/thor/base.rb:444:in `start'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/railties-5.0.0.1/lib/rails/generators.rb:180:in `invoke'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/railties-5.0.0.1/lib/rails/commands/generate.rb:13:in `<top (required)>'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/railties-5.0.0.1/lib/rails/engine/commands_tasks.rb:68:in `require'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/railties-5.0.0.1/lib/rails/engine/commands_tasks.rb:68:in `require_command!'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/railties-5.0.0.1/lib/rails/engine/commands_tasks.rb:73:in `generate_or_destroy'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/railties-5.0.0.1/lib/rails/engine/commands_tasks.rb:44:in `generate'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/railties-5.0.0.1/lib/rails/engine/commands_tasks.rb:37:in `run_command!'
  from /usr/local/opt/rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/railties-5.0.0.1/lib/rails/engine/commands.rb:14:in `<top (required)>'
  from bin/rails:14:in `require'
  from bin/rails:14:in `<main>'
```

Where I am wrong?

File: Rakefile (hunk 2 above)
Comment (bolshakov) [reply Rakefile:10]: Sure, what name do you suggest?

**Comment by bolshakov** [comment]
It is so slow because `'bundle exec rake data:install:migrations RAILS_ENV=test'` takes approximately 1.8 seconds to run. Thus, 8 runs leads to 14.4 total run time. I'm glad to make less such calls, but it requires different approach to testing.

File: lib/tasks/data_migrations.rake (hunk 1 above)
Comment (serggl) [reply lib/tasks/data_migrations.rake:2]: that demo helped me to understand your use case. thank you

**Comment by serggl** [comment]
here is how long it takes with `appraisal rake`
```
real    2m5.188s
user    1m37.169s
sys     0m16.151s
```
a little long...:)

**Comment by bolshakov** [comment]
it should be two times faster now

**Comment by serggl** [comment]
@bolshakov 2 new suggestions:
- instead of copying migration files, maybe we could just run them as a part of `data:migrate` task? So you'll never forget to copy them
- perhaps we can mock `Rails.application.railties` so we will never need to create dummy rails apps and running bundler on it?

**Comment by bolshakov** [comment]
I want to have the similar behaviour with `rake db:install:migration` and `rake db:migrate`. Copying migrations is just an option, not requirement. End user may decide to extend `RailsDataMigrations::Migrator.migrations_path` instead of copying migrations. 

Both approaches are valid and may be used with rails' schema migrations as well.

Despite of mocking `Rails.application.railties` generating real dummy app, makes possible to test against different rails versions. This is the true integration test.

**Comment by vanness145** [comment]
hello

**Comment by bolshakov** [comment]
Hi, @serggl 

> instead of copying migration files, maybe we could just run them as a part of data:migrate task? So you'll never forget to copy them

I tried both approaches, and the main issue with running migrations from an engine is lack of control. For example when you install [Device](https://github.com/plataformatec/devise) migrations, it provides various options what to migrate. And you can disable or rename some of the features. Unfortunately, If you migrate all migrations from engine, you lost all the flexibility. But if you really want to achieve such behaviour, it is easy to do it manually in your engine:

```ruby
class MyEngine < Rails::Engine
  initializer 'install engine migrations' do 
    RailsDataMigrations::Migrator.migrations_path << '../db/data_migrations'
  end
end
```

> perhaps we can mock Rails.application.railties so we will never need to create dummy rails apps and running bundler on it?

I'm a bit out of context now, but it looks like we don't really need the full application at all. I'll try to review PR later.

**Comment by faucct** [comment]
Any chance this could be finished?

**Comment by bolshakov** [comment]
@faucct мне уже сложно вспомнить о чём тут речь, дело было два года назад.

**Pull request closed without merging**